src/translator/swappable.cpp

```diff
@@ -0,0 +1,154 @@
+    engine_->allocator_.free(p);
+  }
+}
+
+void GPULoadedModel::Load(const GPULoadedModel &from) {
+  srcVocabs_ = from.srcVocabs_;
+  trgVocab_ = from.trgVocab_;
+
+  ABORT_IF(engine_ != from.engine_, "TODO: copy across GPUs.");
+
+  for (size_t i = 0; i < parameters_.size(); ++i) {
+    swapper::copyGpuToGpu(reinterpret_cast<char*>(parameters_[i]->data()), reinterpret_cast<const char*>(from.parameters_[i]->data()), parameters_[i]->size(), engine_->myDeviceId_);
+  }
+}
+
+void GPULoadedModel::Load(const CPULoadedModel &from) {
+  srcVocabs_ = from.SrcVocabs();
+  trgVocab_ = from.TrgVocab();
+  for (size_t i = 0; i < parameters_.size(); ++i) {
+    swapper::copyCpuToGpu(reinterpret_cast<char*>(parameters_[i]->data()), from.Parameters()[i].data(), from.Parameters()[i].size(), engine_->myDeviceId_);
+  }
+}
+
+std::string MultilineInputHack(const std::vector<std::string> &input) {
+  if (input.size() == 1) {
+    return input[0];
+  } else {
+    std::string ret;
+    std::size_t size = 0;
+    for (auto&& line : input) {
+      size += line.size() + 1;
+    }
+    ret.reserve(size);
+    for (auto&& line : input) {
+      ret.append(line);
+      ret.append("\n");
+    }
+    return ret;
+  }
+}
+
+Histories GPULoadedModel::Translate(const std::vector<std::string> &input) {
+  ABORT_IF(!trgVocab_, "GPULoadedModel needs to be overwritten by a CPU model first.");
+  engine_->SwapPointers(parameters_);
+
+  auto corpus = New<data::TextInput>(std::vector<std::string>(1, MultilineInputHack(input)), srcVocabs_, engine_->options_); // @TODO dirty hack
+  data::BatchGenerator<data::TextInput> batchGenerator(corpus, engine_->options_, nullptr, false); // @TODO if the asynchronous batch preparation = true, but we supply less text than the mini-batch size we crash
+
+  BeamSearch search(engine_->options_, engine_->scorers_, trgVocab_);
+  Histories ret;
+  ret.reserve(input.size());
+  for (auto&& batch : batchGenerator) {
+    auto result = search.search(engine_->graph_, batch);
+    ret.insert(ret.end(), result.begin(), result.end());
+  }
+  std::sort(ret.begin(), ret.end(),[](marian::Ptr<marian::History> a, marian::Ptr<marian::History> b){return a->getLineNum() < b->getLineNum();});
+  engine_->SwapPointers(parameters_);
+  return ret;
+}
+
+CPULoadedModel::CPULoadedModel(Ptr<Options> options, const std::string &parameters, const std::vector<std::string> &sourceVocabPaths, const std::string &targetVocabPath)
+  : parameters_(io::loadItems(parameters)) {
+  // Load parameters.
+  // Find the special element and remove it:
+  size_t special_idx = 0;
+  for (size_t i = 0; i < parameters_.size(); i++) {
+    if (parameters_[i].name == "special:model.yml") {
+      special_idx = i;
+      break;
+    }
+  }
+  parameters_.erase(parameters_.begin() + special_idx);
+  // Prepare the name so that it matches the named map
+  for (auto&& item : parameters_) {
+    item.name = "F0::" + item.name;
+  }
+  // Sort by name to match params order.
```

Comment:
What "params order" is this referring to? Is it referring to this https://github.com/kpu/marian-dev/pull/2/files#diff-3cf1a9dd6c3dca253c38b1ff4318bc169af8fde7146487efdb5c41c49b12aa72R16 ?

Reply:
Parameters are sorted by name after loading from npz.  

https://github.com/marian-nmt/marian-dev/blob/c29cc83dc4d234f0e3a00a46a729053132b408b8/src/graph/parameters.h#L89-L90